### PR TITLE
[MM-52541] Mark files as deleted along with thread

### PR DESCRIPTION
**Pull request by kostaspt**
#### Summary
When we delete the root post, the thread is deleted as well. The same should happen to files.

#### Ticket Link
https://mattermost.atlassian.net/browse/MM-52541

#### Screenshots
N/A

#### Release Note
```release-note
NONE
```

### CODE REVIEW

File: server/channels/store/sqlstore/post_store.go [hunk 1]
```diff
@@ -2896,6 +2896,35 @@ func (s *SqlPostStore) deleteThread(transaction *sqlxTxWrapper, postId string, d
...
+			From("Posts").
+			Where(sq.Expr("FileInfo.PostId = Posts.Id AND Posts.RootId = ?", postID)).
+			ToSql()
+		if err1 != nil {
+			return errors.Wrapf(err1, "failed to create SQL query to mark files of thread post %s as deleted", postID)
+		}
+
+		_, err = transaction.Exec(queryString, args...)
+	} else {
+		_, err = transaction.NamedExec(`
```
Comment (lieut-data) [inline server/channels/store/sqlstore/post_store.go:2917]: Why `NamedExec` here and query builder above? Seems like we could simplify the error handling (i.e. `err1` juggling) if we just did it consistently?

**Review by lieut-data — CHANGES_REQUESTED** [review]
Nice! Just a comment about how we're building up the query before approving.

File: server/channels/store/sqlstore/post_store.go (hunk 1 above)
Comment (kostaspt) [reply server/channels/store/sqlstore/post_store.go:2917]: Unfortunately, [Squirrel does not support using "join" together with "update"](https://github.com/Masterminds/squirrel/issues/257). Also, MySQL and PostgreSQL do not work in the same way in this case. We could consider using a query with a subquery to collect the IDs that will be compatible with both, but I think this adds complexity as well and possibly a performance hit. Would you happen to have something better in mind?

Comment (lieut-data) [reply server/channels/store/sqlstore/post_store.go:2917]: Oh, right, I'd forgotten about some of the upstream limitations. However, thanks to our fork of squirrel, I think we can do this with multiple tables in the `FROM` statement, e.g.

https://github.com/mattermost/mattermost-server/blob/609ab9c765be14708cfb1d25520cbad7934c325b/server/channels/store/sqlstore/thread_store.go#L634-L652

Comment (kostaspt) [reply server/channels/store/sqlstore/post_store.go]: Oh, I didn't realize we had that change in our fork. Thanks for the heads-up!

**Review by vish9812 — APPROVED** [review]

**Review by esarafianou — APPROVED** [review]
LGTM

**Review by furqanmlk — COMMENTED** [review]
Working as expected with the following steps:

1. Login as syadmin
2. Create  channelA, channelB and userA
3. Post a message in channelA
4. Reply to a message by userA with an attachment
5. Create a permalink of the post with an attachment
6. Paste it into channelB
7. As sysadmin, delete a root post (created on step:03)
8. Go to channelB
9. Click the attachment in the permalink.

Result: attachment is not getting opened.

image

**NOTE**
@kostaspt 
1. The permalink attachment is removed after refreshing the page.
2. Attachment file name is still visible, Is this acceptable for security team?

File: webapp/channels/src/components/post_view/post_message_preview/index.ts [hunk 2]
```diff
@@ -43,7 +43,7 @@ function makeMapStateToProps() {
         let user = null;
         let embedVisible = false;
         let channelDisplayName = ownProps.metadata.channel_display_name;
-        const previewPost = getPost(state, ownProps.metadata.post_id) || ownProps.previewPost;
+        const previewPost = getPost(state, ownProps.metadata.post_id);
```
Comment (larkox) [inline webapp/channels/src/components/post_view/post_message_preview/index.ts:46]: If we are no longer using the previewPost coming from the props, and we are overwriting it, shouldn't we remove it? Also, I am worried on who used to pass the previewPost and how, to make sure they also don't have the post information.

And finally... what if the post is not in the store? (e.g. a post from a channel you haven't navigated to yet)

**Review by larkox — COMMENTED** [review]
I have a few concerns here.

File: webapp/channels/src/components/post_view/post_body_additional_content/post_body_additional_content.tsx [hunk 3]
```diff
@@ -126,7 +126,6 @@ export default class PostBodyAdditionalContent extends React.PureComponent<Props
                 return (
                     <PostMessagePreview
                         metadata={embed.data}
-                        previewPost={embed.data.post}
```
Comment (larkox) [inline webapp/channels/src/components/post_view/post_body_additional_content/post_body_additional_content.tsx:129]: I am a bit worried about were exactly this `embed.data.post` is coming from. Is it stored in the post object in the database? Or is it something that is only in memory? Either way, depending on how this is populated, may still have information about the deleted posts, like the message or the files.

Comment (harshilsharma63) [reply webapp/channels/src/components/post_view/post_body_additional_content/post_body_additional_content.tsx:129]: You're right. We're not clearing this data from post metadata on post deletion. Figuring this out now. Great catch!

File: webapp/channels/src/packages/mattermost-redux/src/reducers/entities/posts.ts [hunk 4]
```diff
@@ -200,11 +202,43 @@ export function handlePosts(state: RelationOneToOne<Post, Post> = {}, action: Ge
...
+                // We'll add everything other than the deleted post's embed here.
+                const newEmbeds: PostEmbed[] = [];
+
+                for (const embed of otherPost.metadata.embeds) {
+                    if (embed.type === 'permalink' && (embed.data as PostPreviewMetadata).post_id === post.id) {
+                        // skip if the embed is the deleted post
+                    } else {
+                        // include everything else
+                        newEmbeds.push(embed);
+                    }
```
Comment (larkox) [inline webapp/channels/src/packages/mattermost-redux/src/reducers/entities/posts.ts:226]: Personal preference, so feel free to keep it as you find it more readable.

```suggestion
                    if (embed.type === 'permalink' && (embed.data as PostPreviewMetadata).post_id === post.id) {
                        // skip if the embed is the deleted post
                        continue;
                    }
                    
                    // include everything else
                    newEmbeds.push(embed);
```

File: webapp/channels/src/packages/mattermost-redux/src/reducers/entities/posts.ts [hunk 5]
```diff
@@ -200,11 +202,43 @@ export function handlePosts(state: RelationOneToOne<Post, Post> = {}, action: Ge
...
+                }
+
+                // if newEmbeds changed, update post's embeds
+                if (newEmbeds.length !== otherPost.metadata.embeds.length) {
+                    // Since otherPost refers to the post from store, its frozen un immutable.
+                    // That's why cloning it here and modifying required parts here this why.
```
Comment (larkox) [inline webapp/channels/src/packages/mattermost-redux/src/reducers/entities/posts.ts:232]:
```suggestion
                    // That's why cloning it and modifying required parts here.
```

**Review by larkox — APPROVED** [review]
LGTM, just two non-blocking nitpicks.

**Comment by yasserfaraazkhan** [comment]
hi @harshilsharma63 @M-ZubairAhmed ,
when the thread is deleted, on spinwick, the screen shows message 'post might have been deleted..' but when upon click `Back to Mattermost` it take user to the `/prepare-workspace` screen

can you check the attachment in the Jira ? (not able to upload screen recording here)
https://mattermost.atlassian.net/browse/MM-52541

**Review by yasserfaraazkhan — APPROVED** [review]
tested it on spin wick.
changes are working as expected.

observation:
- when thread is deleted, the UI still shows `# replies` and `Following` button. This disappeared only when screen is reloaded.

Please let me know if this is to be logged separately or its expected behaviour. 
cc: @lindalumitchell @saturninoabril 
Screenshot 2023-05-30 at 2 03 15 PM

**Comment by harshilsharma63** [comment]
> tested it on spin wick. e2e have passed changes are working as expected.
> 
> observation:
> 
>     * when thread is deleted, the UI still shows `# replies` and `Following` button. This disappears only when the screen is reloaded.
> 
> 
> Please let me know if this is to be logged separately or its expected behaviour. cc: @lindalumitchell @saturninoabril Screenshot 2023-05-30 at 2 03 15 PM

@yasserfaraazkhan this is the existing behavior of deleting a thread and is not changed by this PR. We can update this separately if needed as its independent of this fix.

**Comment by saturninoabril** [comment]
Looking at [MM-T4445 test case](https://github.com/mattermost/mattermost-test-management/blob/main/data/test-cases/channels/collapsed-reply-threads/MM-T4445.md), it should be removed from the global thread once root post is deleted. But I do agree it's a known issue and is [noted in the automated test itself](https://github.com/mattermost/mattermost/blob/master/e2e-tests/cypress/tests/integration/channels/collapsed_reply_threads/global_threads_spec.js#L134-L144). Please file a ticket to address separately.

**Pull request merged**